### PR TITLE
Remove exception filters from Http3RequestStream

### PR DESCRIPTION
**Pull request by steveisok**
I've updated the description as I understand it to be more accurate:

The type of exception filters used cause the linker to generate invalid IL when building the AOT functional tests for iOS.  This in turn causes the mono aot compiler to issue a warning about the invalid IL AND fails the build.

Addresses https://github.com/dotnet/runtime/issues/57010 and unblocks failing builds on CI

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (eerhardt) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:236]: I think we had this same discussion in https://github.com/dotnet/runtime/pull/47197#issuecomment-763607980. The relevant reply copied here:

> Undoing use of exception filters sounds like a step backward to me.

cc @jkotas @stephentoub @marek-safar

Comment (steveisok) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:236]: Yeah, no doubt.  This is intended as an intermediate fix.

Comment (danmoseley) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:236]: Why not revert the change that broke it instead?

Comment (MichalStrehovsky) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:236]: Looking at the IL posted in https://github.com/dotnet/runtime/issues/57010, it looks like illink decided that QuicStreamAbortedException can never be thrown, tried to eliminate this code based on that knowledge, and messed it up.

Adding this somewhere in the method (at the beginning would be a good spot) would likely work around because then illink won't try to touch the filter. It's not smart enough to see this is unreached:

```csharp
#if MONO
if (string.Empty.Length > 0)
    throw new QuicStreamAbortedException("", 0);
#endif
```

Comment (steveisok) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:236]: @MichalStrehovsky Your suggestion didn't seem to work for me.  Is there something else I can try?

**Comment by geoffkizer** [comment]
> Undoing use of exception filters sounds like a step backward to me.

Yes, this.

Is there a plan to make this work with mono AOT?

Is there some sort of guidance about what sort of exception filters work and what ones don't?

I'm having trouble seeing why we are changing legal code to work around underlying platform deficiencies. 

At the very least, if we are going to check this in, we should
(a) Keep the existing code for non mono AOT cases, via an ifdef 
(b) Add a big comment explaining what the hell is going on here, because it's far from obvious

**Comment by steveisok** [comment]
@geoffkizer i updated the PR description to be a little more accurate.  It does not seem to be the aot compiler being unable to handle exception filters.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (MichalStrehovsky) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:236]: If you have the artifacts around, could you attach the System.Net.Http.dll after my workaround? Ideally both the version before/after trimming?

It's possible there are more instances of this in the method body - I want to see if there's any more diffs in the method body after trimming - that would point to all the possible sources of trouble.

Comment (eerhardt) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:236]: > #if MONO

this is System.Net.Http - there is no `MONO` define.

**Comment by eerhardt** [comment]
I think we should move forward with this change in order to unblock CI. But we should get the underlying ILLinker bug fixed in 6.0.

**Comment by stephentoub** [comment]
> I think we should move forward with this change in order to unblock CI. But we should get the underlying ILLinker bug fixed in 6.0.

If we do that, we should prioritize fixing the linker and also revert this change as part of pulling in the fixed linker.  I don't want us to start undoing use of exception filters everywhere.

**Comment by eerhardt** [comment]
> and also revert this change as part of pulling in the fixed linker

+1

> I don't want us to start undoing use of exception filters everywhere.

Agreed. But I don't see a better option. Available options I see are:

0. Let CI be broken until the linker is fixed.
1. The above plan - This change + linker fix + revert this change when possible.
2. Revert #56831 until the linker is fixed

**Comment by stephentoub** [comment]
Is someone working on the linker fix?  What's the expected ETA?

**Comment by eerhardt** [comment]
I pinged @vitek-karas about it this morning. I'm not sure if someone is actively working on it, or an ETA.

**Comment by steveisok** [comment]
> > and also revert this change as part of pulling in the fixed linker
> 
> +1
> 
> > I don't want us to start undoing use of exception filters everywhere.
> 
> Agreed. But I don't see a better option. Available options I see are:
> 
> 1. Let CI be broken until the linker is fixed.
> 2. The above plan - This change + linker fix + revert this change when possible.
> 3. Revert [HTTP/3 Reenabled cookie tests #56831](https://github.com/dotnet/runtime/pull/56831) until the linker is fixed

I forgot #if MONO wasn't defined.  If I take the #if MONO away, @MichalStrehovsky's suggestion seems to work.  I can add a define to the project for mono workloads. Would that be acceptable?

**Comment by ManickaP** [comment]
If this is only induced by #56831, I'd say temporarily disabling the tests is the best option here. At least we do not forget to roll the code back. 

The whole `Http3RequestStream` should be trimmed out by the linker. How come it still plays a role here? Or does the trimming happen afterwards?

And I'm sorry I missed the issue before merging. I've been unwittingly ignoring staging pipeline.

**Review by eerhardt — APPROVED** [review]
Thanks for pushing this forward @steveisok.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (MichalStrehovsky) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs]: > this is System.Net.Http - there is no MONO define.

Sorry for the confusion :(. Glad the workaround works.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs
Comment (lambdageek) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:132]: Should this be `TARGET_MOBILE || TARGET_BROWSER`?

Comment (eerhardt) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs:132]: This file isn't compiled on Browser:

https://github.com/dotnet/runtime/blob/2b3f76c8e3ccb28aefbf5ca879a7cf5933b2642a/src/libraries/System.Net.Http/src/System.Net.Http.csproj#L142-L173

**Pull request merged**
